File: DP/rodCutting.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution{
  public:
  int dp[1005];
  int func(int len,vector<int> &prices)
  {
      if(len==0) return 0;
      int ans=0;
      if(dp[len]!=-1) return dp[len];
      for(int len_to_cut=1;len_to_cut<=len;len_to_cut++)
      {
          if(len>=len_to_cut)
          {
              ans=max(ans,func(len-len_to_cut,prices)+prices[len_to_cut-1]);
          }
      }
      return dp[len]=ans;
      
  }
    int cutRod(int price[], int n) 
    {
        memset(dp,-1,sizeof(dp));
        vector <int>prices(price,price+n);
        return func(n,prices);
    }
};


int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        int a[n];
        for (int i = 0; i < n; i++) 
            cin >> a[i];
            
        Solution ob;

        cout << ob.cutRod(a, n) << endl;
    }
    return 0;
} 

File: Dynamic Programming in java/unbounded knapsack/rod_cutting.java
class Solution{
    public int cutRod(int price[], int n) {
        //code here
        int[] length=new int[n];
        for(int i=0;i<n;i++)
        {
            length[i]=i+1;
        }
        int[][] dp=new int[n+1][n+1];
        for(int i=0;i<n+1;i++)
        {
            for(int j=0;j<n+1;j++)
            {
                if(i==0 && j==0)
                dp[i][j]=1;
                else if(i==0 || j==0)
                dp[i][j]=0;
                else{
                    if(length[i-1]<=j)
                    dp[i][j]=Math.max(price[i-1]+dp[i][j-length[i-1]],dp[i-1][j]);
                    else{
                        dp[i][j]=dp[i-1][j];
                    }
                }
                
            }
        }
        return dp[n][n];
    }
}